File: CodeWars/Mathematics/PerimeterOfSquaresInARectangle/PerimeterOfSquaresInARectangle.rs

struct Fibonacci {
    curr: u64,
    next: u64,
}

impl Fibonacci {
    pub fn new() -> Self {
        Fibonacci { curr: 0, next: 1 }
    }
}

impl Iterator for Fibonacci {
    type Item = u64;

    fn next(&mut self) -> Option<u64> {
        let new_next = self.curr + self.next;

        self.curr = self.next;
        self.next = new_next;

        Some(self.curr)
    }
}

fn perimeter(n: u64) -> u64 {
    Fibonacci::new().take((n+1) as usize).sum::<u64>() * 4
}

fn dotest(n: u64, exp: u64) -> () {
    assert_eq!(perimeter(n), exp)
}

fn main() {
    dotest(5, 80);
    dotest(7, 216);
    dotest(20, 114624);
    dotest(30, 14098308);
}

File: CodeWars/Mathematics/PerimeterOfSquaresInARectangle/PerimeterOfSquaresInARectangle.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef unsigned long long ull;

typedef struct _fib {
    ull curr;
    ull next;
} Fibonacci;

ull fib_next(Fibonacci* fib) {
    ull new_next = fib->curr + fib->next;

    fib->curr = fib->next;
    fib->next = new_next;

    return fib->curr;
}

ull perimeter(int n) {
    Fibonacci fib = { .curr = 0, .next = 1 };
    ull result = 0;

    for (int i = 0 ; i <= n ; ++i)
        result += fib_next(&fib);

    return result*4;
}

void dotest(int n, ull expr) {
    ull act = perimeter(n);
    printf("For numbers: %d Expected: %llu Got: %llu\n", n, expr, act);
    if(act != expr)
        printf("Expected %llu, got %llu\n", expr, act);
    assert(act == expr);
}

int main() {
    {
        dotest(5, 80);
        dotest(7, 216);
        dotest(20, 114624);
        dotest(30, 14098308);
    }
}
